File: python/tests/test_investments_storage.py
import math
import warnings

import pandas as pd

from rts_gmlc_case import data
from rts_gmlc_case.document import CaseDocument, PortfolioCase
from rts_gmlc_case.generation import add_generation
from rts_gmlc_case.investments import data as investments_data
from rts_gmlc_case.investments.storage import BATTERY_TECH_CLASS, add_storage_technologies
from rts_gmlc_case.investments.technologies import GENERATOR_DATABASE, add_supply_technologies
from rts_gmlc_case.investments.topology import add_investment_topology
from rts_gmlc_case.topology import add_topology


def _build(doc: CaseDocument):
    """Build the operations stages (topology + generation, fast: no
    time-series read) plus the investments topology and supply-technology
    stages, then run the storage-technology stage on top -- mirrors
    test_investments_technologies.py's ``_build`` helper.
    """
    op_src = data.rts_source_dir() / "SourceData"
    idx = add_topology(doc, op_src)
    add_generation(doc, op_src, idx)

    portfolio = PortfolioCase(doc, aggregation="Area", base_system_file="system.json")
    inv_src = investments_data.investments_source_dir()
    inv_idx = add_investment_topology(portfolio, inv_src, idx)
    add_supply_technologies(portfolio, inv_src, idx, inv_idx)
    storage_id_by_class = add_storage_technologies(portfolio, inv_src, inv_idx)
    return portfolio, idx, inv_idx, storage_id_by_class


def _battery_row() -> pd.Series:
    gen_db = pd.read_csv(investments_data.investments_source_dir() / GENERATOR_DATABASE)
    rows = gen_db[gen_db["tech"] == BATTERY_TECH_CLASS]
    assert len(rows) == 1
    return rows.iloc[0]


def test_exactly_one_battery_4_row_in_the_generator_database():
    row = _battery_row()
    assert row["GEN UID"] == "313_STORAGE_1"
    assert int(row["Bus ID"]) == 313
    assert float(row["cap"]) == 50.0
    assert float(row["Storage Roundtrip Efficiency"]) == 85.0


def test_both_duration_csvs_have_zero_data_rows():
    # R37: duration_limits is left None on the StorageTechnology because
    # neither duration source has any real data -- this pins that fact
    # against a fresh read so a future dataset refresh that adds rows is
    # caught, not silently left with a stale None.
    source = investments_data.investments_source_dir()
    pshdata = pd.read_csv(source / "storagedata/storage_duration_pshdata.csv")
    storinmaxfrac = pd.read_csv(source / "storagedata/storinmaxfrac.csv")
    assert len(pshdata) == 0
    assert len(storinmaxfrac) == 0


def test_exactly_one_storage_technology_named_battery_4():
    doc = CaseDocument()
    portfolio, idx, inv_idx, storage_id_by_class = _build(doc)

    assert len(portfolio.document.components["StorageTechnology"]) == 1
    assert set(storage_id_by_class) == {BATTERY_TECH_CLASS}

    tech = portfolio.document.components["StorageTechnology"][0]
    assert tech["name"] == BATTERY_TECH_CLASS
    assert tech["id"] == storage_id_by_class[BATTERY_TECH_CLASS]


def test_power_systems_type_is_energy_reservoir_storage_and_present_in_document():
    doc = CaseDocument()
    portfolio, *_ = _build(doc)
    tech = portfolio.document.components["StorageTechnology"][0]

    assert tech["power_systems_type"] == "EnergyReservoirStorage"
    # R39's 4th string, not produced by any of E2's 9 SupplyTechnology
    # classes -- confirmed present with at least one real component here.
    assert tech["power_systems_type"] in doc.document.components
    assert len(doc.document.components[tech["power_systems_type"]]) > 0

    supply_types = {
        supply["power_systems_type"] for supply in portfolio.document.components["SupplyTechnology"]
    }
    assert "EnergyReservoirStorage" not in supply_types


def test_region_is_the_zone_containing_bus_313():
    doc = CaseDocument()
    portfolio, idx, inv_idx, storage_id_by_class = _build(doc)
    tech = portfolio.document.components["StorageTechnology"][0]

    expected_zone_id = inv_idx.zone_id_by_bus_number[313]
    assert tech["region"] == [expected_zone_id]


def test_efficiency_reuses_the_generation_module_formula():
    doc = CaseDocument()
    portfolio, *_ = _build(doc)
    tech = portfolio.document.components["StorageTechnology"][0]

    row = _battery_row()
    expected_leg = math.sqrt(float(row["Storage Roundtrip Efficiency"]) / 100.0)
    assert tech["efficiency"]["in"] == expected_leg
    assert tech["efficiency"]["out"] == expected_leg


def test_storage_tech_and_prime_mover_type():
    doc = CaseDocument()
    portfolio, *_ = _build(doc)
    tech = portfolio.document.components["StorageTechnology"][0]

    assert tech["storage_tech"] == "OTHER_CHEM"
    assert tech["prime_mover_type"] == "BA"


def test_capacity_limits_use_cap_mw_for_power_not_energy():
    doc = CaseDocument()
    portfolio, *_ = _build(doc)
    tech = portfolio.document.components["StorageTechnology"][0]

    row = _battery_row()
    cap_mw = float(row["cap"])
    assert tech["capacity_limits_charge"] == {"min": 0.0, "max": cap_mw}
    assert tech["capacity_limits_discharge"] == {"min": 0.0, "max": cap_mw}
    # cap is a power (MW) figure -- no dataset-backed energy (MWh) capacity
    # figure exists, so capacity_limits_energy is left None, not fabricated.
    assert tech["capacity_limits_energy"] is None


def test_duration_limits_is_none():
    doc = CaseDocument()
    portfolio, *_ = _build(doc)
    tech = portfolio.document.components["StorageTechnology"][0]
    assert tech["duration_limits"] is None


def test_existing_devices_attribute():
    doc = CaseDocument()
    portfolio, idx, inv_idx, storage_id_by_class = _build(doc)

    technology_id = storage_id_by_class[BATTERY_TECH_CLASS]
    existing_devices_by_component_id = {
        assoc.component_id: portfolio.document.supplemental_attributes[i]
        for i, assoc in enumerate(portfolio.document.supplemental_attribute_associations)
        if assoc.attribute_type == "ExistingDevices" and assoc.component_type == "StorageTechnology"
    }
    assert len(existing_devices_by_component_id) == 1
    assert existing_devices_by_component_id[technology_id]["existing_devices"] == [
        "313_STORAGE_1"
    ]


def test_capital_costs_operation_cost_and_financial_data_populated():
    doc = CaseDocument()
    portfolio, *_ = _build(doc)
    tech = portfolio.document.components["StorageTechnology"][0]

    # The three legs travel in one StorageCapitalCost, not as sibling fields.
    assert tech["capital_costs"]["charge_capital_cost"] is not None
    assert tech["capital_costs"]["discharge_capital_cost"] is not None
    assert tech["capital_costs"]["energy_capital_cost"] is not None
    assert tech["operation_costs"] is not None
    assert tech["financial_data"] is not None
    for field in (
        "capital_recovery_period",
        "technology_base_year",
        "debt_fraction",
        "debt_rate",
        "return_on_equity",
        "tax_rate",
    ):
        assert tech["financial_data"][field] is not None


def test_validate_passes():
    doc = CaseDocument()
    portfolio, *_ = _build(doc)
    doc.validate()


def test_no_warnings_building_storage_technologies():
    doc = CaseDocument()
    op_src = data.rts_source_dir() / "SourceData"
    idx = add_topology(doc, op_src)
    add_generation(doc, op_src, idx)
    portfolio = PortfolioCase(doc, aggregation="Area", base_system_file="system.json")
    inv_src = investments_data.investments_source_dir()
    inv_idx = add_investment_topology(portfolio, inv_src, idx)
    add_supply_technologies(portfolio, inv_src, idx, inv_idx)

    with warnings.catch_warnings():
        warnings.simplefilter("error")
        add_storage_technologies(portfolio, inv_src, inv_idx)

File: julia/RTSGMLCCase/test/test_investments_storage.jl
# Build the operations stages this test needs (topology + generation, so the operations
# component types -- ThermalStandard/HydroDispatch/RenewableDispatch/EnergyReservoirStorage --
# are already present in the document), then run the investments topology, supply-technology,
# and storage-technology stages on top -- mirrors test_investments_technologies.jl's
# `_build_supply_technologies` helper.
function _build_storage_technologies(doc)
    op_src = joinpath(RTSGMLCCase.rts_source_dir(), RTSGMLCCase.SOURCE_DATA)
    idx = RTSGMLCCase.add_topology!(doc, op_src)
    RTSGMLCCase.add_generation!(doc, op_src, idx)

    portfolio = RTSGMLCCase.new_portfolio(
        doc;
        aggregation = "Area",
        base_system_file = RTSGMLCCase.SYSTEM_FILENAME,
    )
    inv_src = RTSGMLCCase.investments_source_dir()
    inv_idx = RTSGMLCCase.add_investment_topology!(portfolio, inv_src, idx)
    RTSGMLCCase.add_supply_technologies!(portfolio, inv_src, idx, inv_idx)
    storage_id_by_class = RTSGMLCCase.add_storage_technologies!(portfolio, inv_src, inv_idx)
    return portfolio, idx, inv_idx, storage_id_by_class
end

function _battery_row()
    src = RTSGMLCCase.investments_source_dir()
    gen_db = CSV.read(joinpath(src, RTSGMLCCase.GENERATOR_DATABASE), DataFrame)
    rows = gen_db[gen_db[!, "tech"] .== RTSGMLCCase.BATTERY_TECH_CLASS, :]
    @test nrow(rows) == 1
    return rows[1, :]
end

@testset "exactly one battery_4 row in the generator database" begin
    row = _battery_row()
    @test row["GEN UID"] == "313_STORAGE_1"
    @test Int(row["Bus ID"]) == 313
    @test Float64(row["cap"]) == 50.0
    @test Float64(row["Storage Roundtrip Efficiency"]) == 85.0
end

@testset "both duration CSVs have zero data rows" begin
    # R37: duration_limits is left `nothing` on the StorageTechnology because neither duration
    # source has any real data -- this pins that fact against a fresh read so a future dataset
    # refresh that adds rows is caught, not silently left with a stale `nothing`.
    source = RTSGMLCCase.investments_source_dir()
    pshdata = CSV.read(joinpath(source, "storagedata", "storage_duration_pshdata.csv"), DataFrame)
    storinmaxfrac = CSV.read(joinpath(source, "storagedata", "storinmaxfrac.csv"), DataFrame)
    @test nrow(pshdata) == 0
    @test nrow(storinmaxfrac) == 0
end

@testset "exactly one StorageTechnology named battery_4" begin
    doc = RTSGMLCCase.new_document()
    portfolio, idx, inv_idx, storage_id_by_class = _build_storage_technologies(doc)

    techs = PowerOpenAPIModels.get_components(portfolio.document, "StorageTechnology")
    @test length(techs) == 1
    @test Set(keys(storage_id_by_class)) == Set([RTSGMLCCase.BATTERY_TECH_CLASS])

    tech = only(techs)
    @test tech.name == RTSGMLCCase.BATTERY_TECH_CLASS
    @test tech.id == storage_id_by_class[RTSGMLCCase.BATTERY_TECH_CLASS]
end

@testset "power_systems_type is EnergyReservoirStorage and present in the document" begin
    doc = RTSGMLCCase.new_document()
    portfolio, _ = _build_storage_technologies(doc)
    tech = only(PowerOpenAPIModels.get_components(portfolio.document, "StorageTechnology"))

    @test tech.power_systems_type == "EnergyReservoirStorage"
    # R39's 4th string, not produced by any of E2/JE2's 9 SupplyTechnology classes -- confirmed
    # present with at least one real component here.
    @test length(PowerOpenAPIModels.get_components(doc, tech.power_systems_type)) > 0

    supply_types = Set(s.power_systems_type for s in PowerOpenAPIModels.get_components(portfolio.document, "SupplyTechnology"))
    @test "EnergyReservoirStorage" ∉ supply_types
end

@testset "region is the zone containing bus 313" begin
    doc = RTSGMLCCase.new_document()
    portfolio, idx, inv_idx, storage_id_by_class = _build_storage_technologies(doc)
    tech = only(PowerOpenAPIModels.get_components(portfolio.document, "StorageTechnology"))

    expected_zone_id = inv_idx.zone_id_by_bus_number[313]
    @test tech.region == [expected_zone_id]
end

@testset "efficiency reuses the generation module formula" begin
    doc = RTSGMLCCase.new_document()
    portfolio, _ = _build_storage_technologies(doc)
    tech = only(PowerOpenAPIModels.get_components(portfolio.document, "StorageTechnology"))

    row = _battery_row()
    expected_leg = sqrt(Float64(row["Storage Roundtrip Efficiency"]) / 100.0)
    @test tech.efficiency.in == expected_leg
    @test tech.efficiency.out == expected_leg
end

@testset "storage_tech and prime_mover_type" begin
    doc = RTSGMLCCase.new_document()
    portfolio, _ = _build_storage_technologies(doc)
    tech = only(PowerOpenAPIModels.get_components(portfolio.document, "StorageTechnology"))

    @test tech.storage_tech == PowerOpenAPIModels.StorageTech("OTHER_CHEM")
    @test tech.prime_mover_type == PowerOpenAPIModels.PrimeMovers("BA")
end

@testset "capacity limits use cap MW for power not energy" begin
    doc = RTSGMLCCase.new_document()
    portfolio, _ = _build_storage_technologies(doc)
    tech = only(PowerOpenAPIModels.get_components(portfolio.document, "StorageTechnology"))

    row = _battery_row()
    cap_mw = Float64(row["cap"])
    # Julia wraps the `oneOf` in a generated single-field struct; `.value` unwraps it.
    @test tech.capacity_limits_charge.value.min == 0.0
    @test tech.capacity_limits_charge.value.max == cap_mw
    @test tech.capacity_limits_discharge.value.min == 0.0
    @test tech.capacity_limits_discharge.value.max == cap_mw
    # cap is a power (MW) figure -- no dataset-backed energy (MWh) capacity figure exists, so
    # capacity_limits_energy is left `nothing`, not fabricated.
    @test tech.capacity_limits_energy === nothing
end

@testset "duration_limits is nothing" begin
    doc = RTSGMLCCase.new_document()
    portfolio, _ = _build_storage_technologies(doc)
    tech = only(PowerOpenAPIModels.get_components(portfolio.document, "StorageTechnology"))
    @test tech.duration_limits === nothing
end

@testset "existing devices attribute" begin
    doc = RTSGMLCCase.new_document()
    portfolio, idx, inv_idx, storage_id_by_class = _build_storage_technologies(doc)

    technology_id = storage_id_by_class[RTSGMLCCase.BATTERY_TECH_CLASS]
    existing_devices_by_component_id = Dict(
        assoc.component_id => portfolio.document.supplemental_attributes[i] for
        (i, assoc) in enumerate(portfolio.document.supplemental_attribute_associations) if
        assoc.attribute_type == "ExistingDevices" && assoc.component_type == "StorageTechnology"
    )
    @test length(existing_devices_by_component_id) == 1
    @test existing_devices_by_component_id[technology_id].existing_devices == ["313_STORAGE_1"]
end

@testset "capital costs, operation cost, and financial data populated" begin
    doc = RTSGMLCCase.new_document()
    portfolio, _ = _build_storage_technologies(doc)
    tech = only(PowerOpenAPIModels.get_components(portfolio.document, "StorageTechnology"))

    # The three legs travel in one StorageCapitalCost, not as sibling fields.
    @test tech.capital_costs.charge_capital_cost !== nothing
    @test tech.capital_costs.discharge_capital_cost !== nothing
    @test tech.capital_costs.energy_capital_cost !== nothing
    @test tech.operation_costs !== nothing
    @test tech.financial_data !== nothing
    financial_data = tech.financial_data
    @test financial_data.capital_recovery_period !== nothing
    @test financial_data.technology_base_year !== nothing
    @test financial_data.debt_fraction !== nothing
    @test financial_data.debt_rate !== nothing
    @test financial_data.return_on_equity !== nothing
    @test financial_data.tax_rate !== nothing
end

@testset "validate passes" begin
    doc = RTSGMLCCase.new_document()
    portfolio, _ = _build_storage_technologies(doc)
    PowerOpenAPIModels.validate_document(doc)
end

@testset "expected exactly 1 row precondition raises when battery_4 count differs" begin
    mktempdir() do root
        capdir = joinpath(root, "capacitydata")
        mkpath(capdir)
        src = RTSGMLCCase.investments_source_dir()
        gen_db = CSV.read(joinpath(src, RTSGMLCCase.GENERATOR_DATABASE), DataFrame)
        without_battery = gen_db[gen_db[!, "tech"] .!= RTSGMLCCase.BATTERY_TECH_CLASS, :]
        CSV.write(joinpath(capdir, "ReEDS_generator_database_final_RTS-GMLC_updated_nodal.csv"), without_battery)

        doc = RTSGMLCCase.new_document()
        op_src = joinpath(RTSGMLCCase.rts_source_dir(), RTSGMLCCase.SOURCE_DATA)
        idx = RTSGMLCCase.add_topology!(doc, op_src)
        portfolio = RTSGMLCCase.new_portfolio(
            doc;
            aggregation = "Area",
            base_system_file = RTSGMLCCase.SYSTEM_FILENAME,
        )
        inv_idx = RTSGMLCCase.add_investment_topology!(portfolio, root, idx)

        err = nothing
        try
            RTSGMLCCase.add_storage_technologies!(portfolio, root, inv_idx)
        catch caught
            err = caught
        end
        @test err !== nothing
        message = sprint(showerror, err)
        @test occursin("battery_4", message)
    end
end
